AGPs/AGP-1H3.md

```diff
@@ -0,0 +1,56 @@
+---
+AGP: tbd
+Title: Apiary Activity Feed
+Author: @lkngtn (@1hive)
+Status: Stage II
+Track: Finance
+Created: 2019-11-12
+---
+
+# **AGP-TBD: Apiary Activity Feeds**
+
+## **Address of the transfer recipient**
+
+ `0x9213737fbb72d8f924fca55d508ae2c004413d64` Vault of the [apiary project organization](https://mainnet.aragon.org/#/apiary)
+
+## **Amount of the transfer**
+
+50,000 DAI Due upon approval
+15,000 ANT Due upon successful delivery
+
+## **Number and frequency of transfers if recurring**
+
+2 (One upon approval, one upon successful delivery).
+For the purpose of budgeting, the full amount should be debited against this quarter's budget.
+
+## **Purpose of the transfer**
+
+Due to the work on App Mining, Apiary will be indexing all transactions including "internal interactions". This gives us the ability to analyze **Activities** and we think it would be incredibly useful to be able to surface that information as a contextual activity feed. This feed would make it easier to engage and interact with organizations, making them feel much more like a social network than a company.
+
+The activity feed would be displayed in three contexts within Apiary.
+* On an Organization Detail page, showing all activities related to that organization
+* On a User's profile page, showing all activities involving them across all organizations they participate in.
+* In a User's activity feed, showing all activities across the organization's they have subscribed.   
+
+Activities will be categorized into Primary and Secondary events. Primary actions will be things like `Alice created a Vote`, whereas secondary actions will be things like `Bob voted in favor of option A`. Primary actions will exist as parent elements, and secondary actions will be nested as child elements.
```

Comment:
I'm not sure what you mean by this distinction. Can you elaborate?

Reply:
I think we may want to have some organization and filters to Activities we show in the feed so that they can be organized in a way that makes sense. 

You may have tens or hundreds of castVote events per created vote, but from a user perspective the createVote event is far more important from an informational perspective. So ideally we would want to display the Vote being created in the feed, but then related to that item show the secondary events (each vote being cast). This way you can show cast votes without the entire feed being castVote entries. 

Votes being cast may be the only type of secondary action we have initially, but it feels like an important distinction. 

We also may have some actions that we pick up, but decide not to show in the activity feed. For example trades on Aragon Fundraising would show up as activities, but probably are not super relevant in an activity feed.

Reply:
I'd have to give it a bit more thought on how we could implement this without treating specific transactions as special cases. We might end up having a lot of stuff to maintain because of it, and that would obviously be less than ideal.

Reply:
Hmm, I think we might have to go into it with the expectation that we might have special handling depending on the type of transaction. 

A raw activity feed may be useful in the way that etherscan's internal transaction list is useful, but I think it's important that we are able to remove some of the noise and highlight the most relevant/actionable information.

Reply:
Yes, but there are ways that we could implement it such that it is flexible and up to app developers in the future. For example, Autark might be able to have some control over primary and secondary actions and hidden actions in the activity feed for their apps and so on.

Reply:
If we are able to develop app mining thinking in those activities cases i think that we will be in a good shape to finish app mining and then start with the activity feed

Reply:
> Yes, but there are ways that we could implement it such that it is flexible and up to app developers in the future. For example, Autark might be able to have some control over primary and secondary actions and hidden actions in the activity feed for their apps and so on.

That would be dope, would that come with the tradeoff though of requiring apps to be aware of our API, or would they submit a PR or something to apiary?

Reply:
We could couple it with the "verification" mechanism of app mining if we want to, and app developers would be able to log in to Apiary using Web3 and edit it from there. So, the apps do not need to know about our API and they wouldn't need to submit a PR either. It does sort of lengthen the implementation timeline though

Reply:
I've been checking out how to implement an activity log for the [committees app](https://forum.aragon.org/t/funding-proposal-committees-aragon-app/1327/5) and it's more tricky than I thought. Hopefully our experience can be useful for Apiary's activity feed.

* **"External" transactions** to a contract are relevant for show **user's activity**. Unfortunately they are not indexed. This means that (1) we have to check out all the transactions on the blockchain or (2) we can get all the events on a contract (events are indexed) and check out what transactions originated them. With the second option we may let outside some transactions, so in the case of a software as Apiary that has a blockchain scrapper it has more sense to check out all the transactions.
  * **Apps can forward calls**. In these cases it is relevant what call they are forwarding and to what apps (it can be decoded from their input field, that contains de evmScript).
  * **Forwards doesn't mean the same** in all applications. In the case of a token manager means that an activity has been executed because somebody had the credentials (at least one token of a specific kind), and for voting, it means that a vote has been created. It makes **UI more complicated**.
    - Additionally, **IForwarder** interface **does not provide enough information** to determine if an evmScript has been finally executed: `forward()` can execute the script immediately or store it for later execution, and the only available event when a script is executed is `LogScriptCall` that logs the `msg.sender`, the app that executes the script, and the destination app. It is not enough to do a general case, and we need to complete the picture adding some rules for each app.
* **"Internal" transactions** to a contract are relevant for **contract activity** (as well as external transactions). Internal transactions are not accessible from web3 so scrapping the blockchain is the only way to show a list with all the activities that have happened to an app.
  * **Internal and external transactions may not be enough** to show all the activities, since some of them involve others without calling them explicitly. For example a vote creation can be a vote cast too if a certain parameter is set, and a vote cast can be a vote execution as well if support reaches 100%. In that case knowing the specificities of an application becomes relevant again.
  * Showing the **app's events** can be an alternative to show the app activity log, but we may want to encourage developers to fill up a **radspec description of events** so apiary don't need to hard code what each event of each app stands for.

Reply:
> Internal and external transactions may not be enough to show all the activities, since some of them involve others without calling them explicitly. For example a vote creation can be a vote cast too if a certain parameter is set, and a vote cast can be a vote execution as well if support reaches 100%. In that case knowing the specificities of an application becomes relevant again. 

This is a great point. cc @onbjerg this reinforces the idea that we should expect to have some way to create custom handling for different activity/transaction types.

Reply:
Not entirely sure. This implies that we're not actually looking to track activity but to track every state change in an every app's domain model.

Reply:
> Not entirely sure. This implies that we're not actually looking to track activity but to track every state change in an every app's domain model.

Hmm, from the perspective of the activity feed (as opposed to app mining activity volume), it is sort of the case that what users are likely interested in is understanding and surfacing relevant state changes. 

It does make me wonder about the centralized caching solution you implemented, and whether the activity feed should be based on that model rather than transaction traces?

Reply:
Well it's not something that is actually surfaced in the client UI either when you perform the action unless the Radspec has been formed in a way to include such information, in which case we would definitely catch that in our activity feed. So I'm still not entirely convinced this is the route we want to go.